opm/output/eclipse/Summary.cpp

```diff
@@ -29,223 +29,103 @@
 #include <opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.hpp>
 #include <opm/parser/eclipse/Units/UnitSystem.hpp>
 
+/*
+ * This class takes simulator state and parser-provided information and
+ * orchestrates ert to write simulation results as requested by the SUMMARY
+ * section in eclipse. The implementation is somewhat compact as a lot of the
+ * requested output may be similar-but-not-quite. Through various techniques
+ * the compiler writes a lot of this code for us.
+ */
+
 namespace Opm {
 
 namespace {
 
-/*
- * A series of helper functions to read & compute values from the simulator,
- * intended to clean up the keyword -> action mapping in the *_keyword
- * functions.
- */
-
 using rt = data::Rates::opt;
+using measure = UnitSystem::measure;
 
-/* The supported Eclipse keywords */
-enum class E : out::Summary::kwtype {
-    WBHP,
-    WBHPH,
-    WGIR,
-    WGIRH,
-    WGIT,
-    WGITH,
-    WGOR,
-    WGORH,
-    WGLR,
-    WGLRH,
-    WGPR,
-    WGPRH,
-    WGPT,
-    WGPTH,
-    WLPR,
-    WLPRH,
-    WLPT,
-    WLPTH,
-    WOIR,
-    WOIRH,
-    WOIT,
-    WOITH,
-    WOPR,
-    WOPRH,
-    WOPT,
-    WOPTH,
-    WTHP,
-    WTHPH,
-    WWCT,
-    WWCTH,
-    WWIR,
-    WWIRH,
-    WWIT,
-    WWITH,
-    WWPR,
-    WWPRH,
-    WWPT,
-    WWPTH,
-    GWPT,
-    GOPT,
-    GGPT,
-    GWPR,
-    GOPR,
-    GLPR,
-    GGPR,
-    GWIR,
-    GWIT,
-    GGIR,
-    GGIT,
-    GGIRH,
-    GWIRH,
-    GOIRH,
-    GGITH,
-    GWITH,
-    GWPRH,
-    GOPRH,
-    GGPRH,
-    GLPRH,
-    GWPTH,
-    GOPTH,
-    GGPTH,
-    GLPTH,
-    GWCT,
-    GGOR,
-    UNSUPPORTED,
-};
+/* Some numerical value with its unit tag embedded to enable caller to apply
+ * unit conversion. This removes a ton of boilerplate.
+ *
+ * Some arithmetic operations are supported, which also codifies the following
+ * assumptions:
+ * * Division by zero, such as in gas/oil ratios without oil, shall default to 0
+ * * Division typically happen to calculate ratios, in which case unit
+ *   conversion is a no-op
+ * * When performing an operation on two operands, the left operand's unit is
+ *   applied, meaning arithmetic is **not** associative with respect to units.
```

Comment:
1. The arithmetic is, it's the type information that behaves possibly unexpectedly. Our unit system is not complete enough to have reasonable type arithmetic (I really wish we had, see `boost-units` for a cool example), but even then it would be quirky.

Consider WGOR, well gas-oil ratio, that I'm pretty sure use different units (!!) for rates. The computation is G/O, but the result is a ratio and unitless, but unless `/` erases that then unit conversion will be applied later. Another nice one is `duration * rate` which results in volume. This **could** have been handled by unit-aware multiplication, but isn't as of now because we don't have any resemblance of support. This system takes us somewhat closer, but it does unfortunately introduce this detail.

2. I honestly don't know what eclipse does, but x/0 surely does happen naturally and returning 0 is probably better than arithmetic exceptions. Example is WGOR as above, when a well produce no oil (or switch to an injector). Regardless, the output would be garbage.